### PR TITLE
introduce struct BasicProperties for basicPublish

### PR DESCRIPTION
**Pull request by bollhals**
## Proposed Changes

Introduces a struct BasicProperties and opens up the header implementation to a public.

Changes in detail:
- Replaced IBasicProperties in the receiving area with ReadOnlyBasicProperties
- Replaced BasicProperties class with a struct version
- Modified BasicPublish to be generic (TProperties : IBasicProperties, IAmqpHeader)
- Removed CreateBasicProperties method, replacement is 'new BasicProperties()'
- Introduce public IAmqpHeader interface. (Can be implemented by anyone and used as header properties)
- Removed IStreamProperties, IContentHeader (Not needed for anything)
- Aligned the name for getting the buffer size: GetRequiredPayloadBufferSize -> GetRequiredBufferSize
- Aligned the name for writing to the buffer: WritePropertiesTo -> WriteTo

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in related repositories

## Further Comments

I'm going to post a few benchmark numbers as soon as I get around to run them. (soon)

### CODE REVIEW

**Comment by michaelklishin** [comment]
Thank you. Looks like this is still a WIP? I run into compilation errors:

```
/Users/antares/Development/RabbitMQ/dotnet_client.git/projects/Unit/TestConnectionRecovery.cs(783,20): error CS0411: The type arguments for method 'IModel.BasicPublish<TProperties>(string, string, in TProperties, ReadOnlyMemory<byte>, bool)' cannot be inferred from the usage. Try specifying the type arguments explicitly. [/path/to/dotnet_client.git/projects/Unit/Unit.csproj]
/path/to/dotnet_client.git/projects/Unit/TestConnectionRecovery.cs(798,37): error CS1061: 'IModel' does not contain a definition for 'CreateBasicProperties' and no accessible extension method 'CreateBasicProperties' accepting a first argument of type 'IModel' could be found (are you missing a using directive or an assembly reference?) [/path/to/dotnet_client.git/projects/Unit/Unit.csproj]
```

**Comment by bollhals** [comment]
Ah probably my last rebase against main... sorry for that. I'll update it when I ran the benchmarks👍🏼

**Comment by bollhals** [comment]
|                    Method | When   |            Runtime |      Mean |    Error |   StdDev | Code Size |
|-------------------------- |--------|------------------- |----------:|---------:|---------:|----------:|
| BasicPublishWriteNonEmpty | Before |      .NET Core 3.1 | 324.15 ns | 2.803 ns | 2.622 ns |   5,638 B |
| BasicPublishWriteNonEmpty | After  |      .NET Core 3.1 | 327.46 ns | 2.391 ns | 2.236 ns |   5,697 B |
|         BasicPublishWrite | Before |      .NET Core 3.1 | 118.49 ns | 1.718 ns | 1.607 ns |   5,638 B |
|         BasicPublishWrite | After  |      .NET Core 3.1 |  86.68 ns | 0.507 ns | 0.450 ns |   2,651 B |
|   BasicPublishMemoryWrite | Before |      .NET Core 3.1 |  72.80 ns | 0.386 ns | 0.342 ns |   6,127 B |
|   BasicPublishMemoryWrite | After  |      .NET Core 3.1 |  44.78 ns | 0.288 ns | 0.270 ns |   2,594 B |
| BasicPublishWriteNonEmpty | Before | .NET Framework 4.8 | 438.34 ns | 3.226 ns | 3.018 ns |   6,831 B |
| BasicPublishWriteNonEmpty | After  | .NET Framework 4.8 | 443.86 ns | 4.187 ns | 3.712 ns |   6,849 B |
|         BasicPublishWrite | Before | .NET Framework 4.8 | 220.12 ns | 0.528 ns | 0.468 ns |   6,703 B |
|         BasicPublishWrite | After  | .NET Framework 4.8 | 182.63 ns | 0.969 ns | 0.906 ns |   3,006 B |
|   BasicPublishMemoryWrite | Before | .NET Framework 4.8 | 153.67 ns | 1.525 ns | 1.427 ns |   7,044 B |
|   BasicPublishMemoryWrite | After  | .NET Framework 4.8 | 119.55 ns | 0.652 ns | 0.610 ns |   3,347 B |

|              Method | When |     Mean |   Error |  StdDev | Ratio |    Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------- |------|---------:|--------:|--------:|------:|---------:|------:|------:|----------:|
| Publish_Hello_World | Before |149.2 ms | 2.78 ms | 2.60 ms |  1.00 | 750.0000 |     - |     - |   4.33 MB |
| Publish_Hello_World | After | 148.1 ms | 2.94 ms | 4.66 ms |  1.00 | 750.0000 |     - |     - |   4.02 MB |

Before
Before

After
after


Summary:
Pure serialization performance is slight worse, but overall the performance is better and there is less memory allocations.
Also performance without properties is significant faster.

Remaining memory allocation for sending and receiving is mostly string creation.

**Comment by bollhals** [comment]
rebased after latest additions in main.

**Comment by bollhals** [comment]
can this be reviewed?

**Comment by bollhals** [comment]
Reminder for review

**Comment by michaelklishin** [comment]
I had to update `RabbitMQ.Client.Unit.APIApproval` expectations, running the rest of the suites.

**Comment by bollhals** [comment]
Whoops, my idea was just to review, as this change does have impact on some basic API that will require some turtorials/docu update. 

Were you aware of this?

File: projects/RabbitMQ.Client/client/impl/CommandAssembler.cs [hunk 1]
```diff
@@ -128,28 +128,32 @@ private void ParseMethodFrame(in InboundFrame frame)
...
+        private bool ParseHeaderFrame(in InboundFrame frame)
         {
             if (frame.Type != FrameType.FrameHeader)
             {
                 throw new UnexpectedFrameException(frame.Type);
             }
 
             ReadOnlySpan<byte> span = frame.Payload.Span;
-            _header = _protocol.DecodeContentHeaderFrom(NetworkOrderDeserializer.ReadUInt16(span), span.Slice(12));
+            var classId =  NetworkOrderDeserializer.ReadUInt16(span);
```
Comment (stebet) [inline projects/RabbitMQ.Client/client/impl/CommandAssembler.cs:139]: Ahh, this is where #1260 was introduced :)

**Pull request merged**